File: backend/app/services/shop_service.py
from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from app.repositories.shop_repo import shop_repo
from app.repositories.product_repo import product_repo
from app.schemas.shop import ShopCreate, ShopUpdate
from app.schemas.product import ProductCreate
from app.models.user import UserRole, User
from app.models.shop import Shop
from app.models.product import Product

class ShopService:
    async def create_shop(self, db: AsyncSession, shop_in: ShopCreate, current_user: User) -> Shop:
        """Register a new shop under a certified Shop Owner."""
        if current_user.role != UserRole.SHOP_OWNER.value and not current_user.is_admin:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Only registered Shop Owners can open a shop."
            )
        
        obj_in = {
            "name": shop_in.name,
            "description": shop_in.description,
            "logo_url": shop_in.logo_url,
            "owner_id": current_user.id
        }
        return await shop_repo.create(db, obj_in=obj_in)

    async def add_product_to_shop(self, db: AsyncSession, shop_id: str, product_in: ProductCreate, current_user: User) -> Product:
        """Add a new product catalog listing linked to the owner's shop."""
        shop = await shop_repo.get(db, shop_id)
        if not shop:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Shop not found."
            )
        
        # Verify ownership
        if shop.owner_id != current_user.id and not current_user.is_admin:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You do not own this shop."
            )

        obj_in = product_in.model_dump()
        obj_in["shop_id"] = shop_id
        return await product_repo.create(db, obj_in=obj_in)

shop_service = ShopService()

File: backend/src/services/shopService.js
import { v4 as uuidv4 } from 'uuid';
import supabase from '../db/supabase.js';
import { AppError } from '../middleware/error.js';

export const shopService = {
  async createShop(shopData, currentUser) {
    if (currentUser.role !== 'shop_owner' && !currentUser.is_admin) {
      throw new AppError(403, 'Only registered Shop Owners can open a shop.');
    }

    const shopId = uuidv4();
    const { data: shop, error } = await supabase
      .from('shops')
      .insert({
        id: shopId,
        owner_id: currentUser.id,
        name: shopData.name,
        description: shopData.description,
        logo_url: shopData.logo_url,
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString()
      })
      .select('*')
      .single();

    if (error) {
      throw new AppError(500, `Error opening shop: ${error.message}`);
    }

    return shop;
  },

  async addProductToShop(shopId, productData, currentUser) {
    // 1. Fetch shop details
    const { data: shop, error: shopError } = await supabase
      .from('shops')
      .select('*')
      .eq('id', shopId)
      .single();

    if (shopError || !shop) {
      throw new AppError(404, 'Shop not found.');
    }

    // 2. Verify shop ownership
    if (shop.owner_id !== currentUser.id && !currentUser.is_admin) {
      throw new AppError(403, 'You do not own this shop.');
    }

    // 3. Insert product
    const productId = uuidv4();
    const { data: product, error: productError } = await supabase
      .from('products')
      .insert({
        id: productId,
        shop_id: shopId,
        name: productData.name,
        description: productData.description,
        price: productData.price,
        original_price: productData.original_price,
        stock: productData.stock,
        image_url: productData.image_url,
        images: productData.images,
        category: productData.category,
        subcategory: productData.subcategory,
        brand: productData.brand,
        sku: productData.sku,
        weight: productData.weight,
        dimensions: productData.dimensions,
        rating: 0.0,
        review_count: 0,
        is_featured: productData.is_featured || false,
        is_active: productData.is_active !== undefined ? productData.is_active : true,
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString()
      })
      .select('*')
      .single();

    if (productError) {
      throw new AppError(500, `Error listing product to shop: ${productError.message}`);
    }

    return product;
  }
};

export default shopService;
